Source/Core/VideoBackends/D3D12/D3DUtil.cpp

```diff
@@ -0,0 +1,1051 @@
+// Copyright 2010 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cctype>
+#include <list>
+#include <string>
+
+#include "VideoBackends/D3D12/D3DBase.h"
+#include "VideoBackends/D3D12/D3DCommandListManager.h"
+#include "VideoBackends/D3D12/D3DDescriptorHeapManager.h"
+#include "VideoBackends/D3D12/D3DShader.h"
+#include "VideoBackends/D3D12/D3DState.h"
+#include "VideoBackends/D3D12/D3DTexture.h"
+#include "VideoBackends/D3D12/D3DUtil.h"
+
+#include "VideoBackends/D3D12/Render.h"
+#include "VideoBackends/D3D12/StaticShaderCache.h"
+
+namespace DX12
+{
+
+extern D3D12_BLEND_DESC g_reset_blend_desc;
+extern D3D12_DEPTH_STENCIL_DESC g_reset_depth_desc;
+extern D3D12_RASTERIZER_DESC g_reset_rast_desc;
+
+namespace D3D
+{
+
+void ResourceBarrier(ID3D12GraphicsCommandList* command_list, ID3D12Resource* resource, D3D12_RESOURCE_STATES state_before, D3D12_RESOURCE_STATES state_after, UINT subresource)
+{
+	if (state_before == state_after)
+		return;
+
+	CHECK(resource, "NULL resource passed to ResourceBarrier.");
+
+	D3D12_RESOURCE_BARRIER resourceBarrierDesc = {
+		D3D12_RESOURCE_BARRIER_TYPE_TRANSITION, // D3D12_RESOURCE_TRANSITION_BARRIER_DESC Transition
+		D3D12_RESOURCE_BARRIER_FLAG_NONE,       // D3D12_RESOURCE_BARRIER_FLAGS Flags
+
+		// D3D12_RESOURCE_TRANSITION_BARRIER_DESC Transition
+		{
+			resource,    // ID3D12Resource *pResource;
+			subresource, // UINT Subresource;
+			state_before, // UINT StateBefore;
+			state_after   // UINT StateAfter;
+		}
+	};
+
+	command_list->ResourceBarrier(1, &resourceBarrierDesc);
+}
+
+// Ring buffer class, shared between the draw* functions
+class UtilVertexBuffer
+{
+public:
+	explicit UtilVertexBuffer(int size) : m_max_size(size)
+	{
+		CheckHR(
+			device12->CreateCommittedResource(
+				&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_UPLOAD),
+				D3D12_HEAP_FLAG_NONE,
+				&CD3DX12_RESOURCE_DESC::Buffer(m_max_size),
+				D3D12_RESOURCE_STATE_GENERIC_READ,
+				nullptr,
+				IID_PPV_ARGS(&m_buf12)
+				)
+			);
+
+		// Map buffer for CPU access upon creation.
+		// On D3D12, CPU-visible resources are persistently mapped.
+
+		CheckHR(m_buf12->Map(0, nullptr, &m_buf12_data));
+	}
+	~UtilVertexBuffer()
+	{
+		D3D::command_list_mgr->DestroyResourceAfterCurrentCommandListExecuted(m_buf12);
```

Comment:
For upload heaps, it is ok to leave them persistently mapped, unmapping doesn't really do anything AFAIK: https://msdn.microsoft.com/en-us/library/windows/desktop/dn788712(v=vs.85).aspx